src/bin/pg_autoctl/monitor.c

```diff
@@ -1881,6 +1945,13 @@ monitor_listen_loop(Monitor *monitor)
 	 */
 	for (;;)
 	{
+		if (!ensure_monitor_pg_running(monitor, mconfig))
+		{
+			log_warn("Failed to ensure PostgreSQL is running, retrying in %d seconds", PG_AUTOCTL_MONITOR_SLEEP_TIME);
+			sleep(PG_AUTOCTL_MONITOR_SLEEP_TIME);
+			continue;
+		}
```

Comment:
We still have the problem that when we lose the connection, we do `pgsql_listen` before we ensure that Postgres is running. Maybe your initial placement of the function call was right, or maybe we need to re-think that loop and make it a `do { ... } while (true)` one instead, where it might be easier to reason and include the first step in the main loop.

Reply:
ensure_monitor_pg_running issues `pgsql_finish` if the db is known to be down. This way the first call to  `monitor_get_notifications` will fail, and issue `pgsql_listen` again. I don't see any benefit to move `monitor_get_notifications` before `ensure_monitor_pg_running`